hyppo/kgof/_utils.py

```diff
@@ -0,0 +1,310 @@
+        e = self.end
+        c = self.chunk_size
+        # Probably not a good idea to use list. Waste memory.
+        L = list(range(s, e, c))
+        L.append(e)
+        return zip(L, L[1:])
+
+# end ChunkIterable
+
+def constrain(val, min_val, max_val):
+    return min(max_val, max(min_val, val))
+
+def dist_matrix(X, Y):
+    """
+    Construct a pairwise Euclidean distance matrix of size X.shape[0] x Y.shape[0]
+    """
+    sx = np.sum(X**2, 1)
+    sy = np.sum(Y**2, 1)
+    D2 =  sx[:, np.newaxis] - 2.0*X.dot(Y.T) + sy[np.newaxis, :] 
+    # to prevent numerical errors from taking sqrt of negative numbers
+    D2[D2 < 0] = 0
+    D = np.sqrt(D2)
+    return D
+
+def dist2_matrix(X, Y):
+    """
+    Construct a pairwise Euclidean distance **squared** matrix of size
+    X.shape[0] x Y.shape[0]
+    """
+    sx = np.sum(X**2, 1)
+    sy = np.sum(Y**2, 1)
+    D2 =  sx[:, np.newaxis] - 2.0*np.dot(X, Y.T) + sy[np.newaxis, :] 
+    return D2
+
+def meddistance(X, subsample=None, mean_on_fail=True):
+    """
+    Compute the median of pairwise distances (not distance squared) of points
+    in the matrix.  Useful as a heuristic for setting Gaussian kernel's width.
+    Parameters
+    ----------
+    X : n x d numpy array
+    mean_on_fail: True/False. If True, use the mean when the median distance is 0.
+        This can happen especially, when the data are discrete e.g., 0/1, and 
+        there are more slightly more 0 than 1. In this case, the m
+    Return
+    ------
+    median distance
+    """
+    if subsample is None:
+        D = dist_matrix(X, X)
+        Itri = np.tril_indices(D.shape[0], -1)
+        Tri = D[Itri]
+        med = np.median(Tri)
+        if med <= 0:
+            # use the mean
+            return np.mean(Tri)
+        return med
+
+    else:
+        assert subsample > 0
+        rand_state = np.random.get_state()
+        np.random.seed(9827)
+        n = X.shape[0]
+        ind = np.random.choice(n, min(subsample, n), replace=False)
+        np.random.set_state(rand_state)
+        # recursion just one
+        return meddistance(X[ind, :], None, mean_on_fail)
+
+
+def is_real_num(X):
+    """return true if x is a real number. 
+    Work for a numpy array as well. Return an array of the same dimension."""
+    def each_elem_true(x):
+        try:
+            float(x)
+            return not (np.isnan(x) or np.isinf(x))
+        except:
+            return False
+    f = np.vectorize(each_elem_true)
+    return f(X)
+    
+
+def tr_te_indices(n, tr_proportion, seed=9282 ):
+    """Get two logical vectors for indexing train/test points.
+    Return (tr_ind, te_ind)
+    """
+    rand_state = np.random.get_state()
+    np.random.seed(seed)
+
+    Itr = np.zeros(n, dtype=bool)
+    tr_ind = np.random.choice(n, int(tr_proportion*n), replace=False)
+    Itr[tr_ind] = True
+    Ite = np.logical_not(Itr)
+
+    np.random.set_state(rand_state)
+    return (Itr, Ite)
+
+def subsample_ind(n, k, seed=32):
```

Comment:
Couldn't find anything that suits this exact purpose.

Reply:
Just use `np.random.choice` directly with a `rng` instance like mentioned before

Reply:
That's what the function does, just wraps it in a function to make it easier to use since I'm using it over and over again.`
    rng = default_rng(seed);
    ind = rng.choice(n, k, replace=False);
    return ind;
`

Reply:
sure, but if you need, you can define your `rng` as a class attribute. So your reducing the 10-15 lines of code for the function and classes (including `NumpySeedContext`) into 2 lines

Reply:
`NumpySeedContext` has since been removed, and I replaced it everywhere it showed up throughout the module with an `rng` instance. So in this case, rather than going back and changing all my changes, should I just replace any usage of `subsample_ind` in the module with `rng` instances to maintain consistency?

Reply:
yes

Reply:
Removed and replaced!